packages/contracts-bedrock/src/L2/L1BlockHolocene.sol

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+import { L1Block, ConfigType } from "src/L2/L1Block.sol";
+import { StaticConfig } from "src/libraries/StaticConfig.sol";
+import "src/libraries/L1BlockErrors.sol";
+
+/// @custom:proxied
+/// @custom:predeploy 0x4200000000000000000000000000000000000015
+/// @title L1BlockHolocene
+/// @notice Holocene extenstions of L1Block.
+contract L1BlockHolocene is L1Block {
+    /// @notice Event emitted when a new batcher hash is set.
+    event BatcherHashSet(bytes32 indexed batcherHash);
+
+    /// @notice Event emitted when new fee scalars are set.
+    event FeeScalarsSet(uint32 indexed blobBasefeeScalar, uint32 indexed basefeeScalar);
+
+    /// @notice Event emitted when a new gas limit is set.
+    event GasLimitSet(uint64 indexed gasLimit);
+
+    /// @notice The gas limit of L2 blocks in the same epoch.
+    uint64 public gasLimit;
+
+    /// @notice Sets static configuration options for the L2 system. Can only be called by the special
+    ///         depositor account.
+    /// @param _type  The type of configuration to set.
+    /// @param _value The encoded value with which to set the configuration.
+    function setConfig(ConfigType _type, bytes calldata _value) public override {
+        if (msg.sender != DEPOSITOR_ACCOUNT()) revert NotDepositor();
+
+        if (_type == ConfigType.SET_BATCHER_HASH) {
+            _setBatcherHash(_value);
+        } else if (_type == ConfigType.SET_FEE_SCALARS) {
+            _setFeeScalars(_value);
+        } else if (_type == ConfigType.SET_GAS_LIMIT) {
+            _setGasLimit(_value);
+        } else {
+            super.setConfig(_type, _value);
+        }
+    }
+
+    /// @notice Internal method to set new batcher hash.
+    /// @param _value The encoded value with which to set the new batcher hash.
+    function _setBatcherHash(bytes calldata _value) internal {
+        bytes32 _batcherHash = StaticConfig.decodeSetBatcherHash(_value);
+
+        batcherHash = _batcherHash;
+
+        emit BatcherHashSet(_batcherHash);
+    }
+
+    /// @notice Internal method to set new fee scalars.
+    /// @param _value The encoded value with which to set the new fee scalars.
+    function _setFeeScalars(bytes calldata _value) internal {
+        uint256 _scalar = StaticConfig.decodeSetFeeScalars(_value);
+
+        (uint32 _blobBasefeeScalar, uint32 _basefeeScalar) = _decodeScalar(_scalar);
+
+        blobBaseFeeScalar = _blobBasefeeScalar;
+        baseFeeScalar = _basefeeScalar;
+
+        emit FeeScalarsSet(_blobBasefeeScalar, _basefeeScalar);
+    }
+
+    /// @notice Internal method to decode blobBaseFeeScalar and baseFeeScalar.
+    /// @return Decoded blobBaseFeeScalar and baseFeeScalar.
+    function _decodeScalar(uint256 _scalar) internal pure returns (uint32, uint32) {
```

Comment:
I think we will want encode/decode logic in a library so it can be a single source of truth and then imported everywhere that it is required so we only implement the logic once and we can unit test roundtrip

Reply:
I completely agree with your point. However, there's a nuance we need to consider: `StaticConfig.encodeXXX` returns `bytes`, while we need to maintain the [scalar](https://github.com/ethereum-optimism/optimism/blob/d7e5418745ef5da2df999a98da11cde228a4fbaa/packages/contracts-bedrock/src/L1/SystemConfigHolocene.sol#L47) variable as uint256. This is the primary reason it's not encapsulated within the StaticConfig library.
I understand the desire to undo the packing hack, but we face a constraint. We still need to maintain the [scalar](https://github.com/ethereum-optimism/optimism/blob/d7e5418745ef5da2df999a98da11cde228a4fbaa/packages/contracts-bedrock/src/L1/SystemConfigHolocene.sol#L47) variable for backwards compatibility. Consequently, we can't undo the packing hack unless we're prepared to discontinue backwards compatibility for the scalar variable.

Reply:
Good point, the packing logic are now moved into the library in this commit: https://github.com/ethereum-optimism/optimism/pull/11600/commits/e876fe4029ddac58df548661fdc892267b975004